### PR TITLE
Avoid needless map copy during disk usage simulation

### PR DESCRIPTION
**Pull request by idegtiarenko**
Currently, we have to copy the `shardSizes` map in disk usage simulation 
during desired balance computation. The copy is performed twice (cluster 
info > simulator & simulator > cluster info) for each computation iteration. 
The map is generally big, its size is proportional to the number of shards in 
cluster. This is wasteful as we only modify imported map in case we have new 
shards starting. This change makes the copying lazy. 

This is my experiment of rebalancing a big cluster where all shards are started `copyOf` is responsible for 6% of samples and is eliminated completely with this change.

Before:
before

After:
after

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/common/util/LazyCopyOnWriteMap.java [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+
+/**
+ * This map is designed to be constructed from an immutable map and only be copied if a (rare) mutation operation occurs.
+ * It could be converted back to an immutable map using `org.elasticsearch.common.util.LazyCopyOnWriteMap#toImmutableMap()`.
+ */
+public class LazyCopyOnWriteMap<K, V> implements Map<K, V> {
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/common/util/LazyCopyOnWriteMap.java:25]: Naming nit/suggestion: no need to call this `Lazy` but maybe `CopyOnWriteMap` implies too much thread-safety.

```suggestion
public class CopyOnFirstWriteMap<K, V> implements Map<K, V> {
```

**Review by DaveCTurner — COMMENTED** [review]
Makes sense - I left a couple of suggestions.

File: server/src/main/java/org/elasticsearch/common/util/LazyCopyOnWriteMap.java [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+
+    public LazyCopyOnWriteMap(Map<K, V> source) {
+        this.source = source;
+        this.copy = null;
+    }
+
+    private Map<K, V> getForUpdate() {
+        if (copy == null) {
+            this.copy = new HashMap<>(source);
+            this.source = null;
```
[the remark this answers is not in the document]
Comment (idegtiarenko) [reply server/src/main/java/org/elasticsearch/common/util/LazyCopyOnWriteMap.java:40]: I believe it is cleaner this way: it is easy to understand if this was ever modified.

**Review by DaveCTurner — APPROVED** [review]
LGTM

**Pull request merged**